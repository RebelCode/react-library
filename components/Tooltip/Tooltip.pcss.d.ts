export const root: string
export const container: string
export const containerTop: string
export const containerBottom: string
export const containerLeft: string
export const containerRight: string
export const content: string
export const arrow: string
export const arrowTop: string
export const arrowBottom: string
export const arrowLeft: string
export const arrowRight: string
interface Namespace {
	"root": string,
	"container": string,
	"containerTop": string,
	"container-top": string,
	"containerBottom": string,
	"container-bottom": string,
	"containerLeft": string,
	"container-left": string,
	"containerRight": string,
	"container-right": string,
	"content": string,
	"arrow": string,
	"arrowTop": string,
	"arrow-top": string,
	"arrowBottom": string,
	"arrow-bottom": string,
	"arrowLeft": string,
	"arrow-left": string,
	"arrowRight": string,
	"arrow-right": string,
}
declare const tooltipPcss: Namespace
export default tooltipPcss
