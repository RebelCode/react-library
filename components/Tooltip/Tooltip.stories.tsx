import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import Tooltip from "./index";

export default {
  title: "Components/Tooltip",
  component: Tooltip,
  parameters: {
    // More on Story layout: https://storybook.js.org/docs/react/configure/story-layout
    layout: "fullscreen",
  },
} as ComponentMeta<typeof Tooltip>;

const Template: ComponentStory<typeof Tooltip> = (args) => (
  <Tooltip {...args} />
);

export const Initial = Template.bind({});
Initial.args = {
  visible: true,
  children: [()=><button>Submit</button>, <h1>sdlkmsldkksm</h1>]
};
