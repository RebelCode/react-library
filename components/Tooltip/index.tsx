import css from "./Tooltip.pcss"
import React, {useEffect} from "react"
import ReactDOM from "react-dom"
import {Manager, Popper, Reference} from "react-popper"
import {DualPopperChildren, Placement} from "popper"
import {classList} from "../../utils/classes"

type ThemeProp = {
  root?: string,
  content?: string,
  container?: string,
  containerTop?: string,
  containerBottom?: string,
  containerLeft?: string,
  containerRight?: string;
  arrow?: string,
  arrowTop?: string,
  arrowBottom?: string,
  arrowLeft?: string,
  arrowRight?: string,
}

type Props = {
  visible: boolean;
  delay?: number;
  children: DualPopperChildren;
  placement?: Placement;
  theme?: ThemeProp;
}

export default function Tooltip({visible, delay, placement, theme, children}: Props) {
  theme = theme ?? {}
  placement = !placement ? "bottom" : placement

  const [showTooltip, setShowTooltip] = React.useState(false)
  const modifiers = {
    preventOverflow: {
      boundariesElement: document.body,
      padding: 5,
    },
  }

  useEffect(() => {
    const ms = visible ? delay : 1
    const timeout = setTimeout(() => setShowTooltip(visible), ms)

    return () => clearTimeout(timeout)
  }, [visible])

  const containerClassKey = getAlignedClass("container", placement)
  const arrowClassKey = getAlignedClass("arrow", placement)

  const containerClass = classList(
    (css as any)[containerClassKey],
    theme.container as string,
    (theme as any)[containerClassKey],
  )
  const arrowClass = classList(
    (css as any)[arrowClassKey],
    theme.arrow as string,
    (theme as any)[arrowClassKey],
  )

  return (
    <Manager aria-label="Tooltip">
      <Reference>
        {(props) => children[0](props)}
      </Reference>
      <Popper placement={placement} modifiers={modifiers} positionFixed={true} aria-label="Tooltip">
        {
          ({ref, style, placement, arrowProps}) => !showTooltip ? null : (
            ReactDOM.createPortal(
              <div
                ref={ref}
                className={classList(css.root, theme!.root as string)}
                style={style}
                aria-label="Tooltip"
                tabIndex={-1}>
                <div className={containerClass} data-placement={placement} aria-label="Tooltip" >
                  <div className={classList(css.content, theme!.content as string)}>
                    {children[1]}
                  </div>
                  <div
                    className={arrowClass}
                    ref={arrowProps.ref}
                    style={arrowProps.style}
                    data-placement={placement} aria-label="Tooltip"/>
                </div>
              </div>,
              document.body,
            )
          )
        }
      </Popper>
    </Manager>
  )
};

function getAlignedClass(className: string, placement: Placement) {
  switch (placement) {
    case "top":
    case "top-start":
    case "top-end":
      return className + "Top"
    case "bottom":
    case "bottom-start":
    case "bottom-end":
      return className + "Bottom"
    case "left":
    case "left-start":
    case "left-end":
      return className + "Left"
    case "right":
    case "right-start":
    case "right-end":
      return className + "Right"
    default:
      return className
  }
}
