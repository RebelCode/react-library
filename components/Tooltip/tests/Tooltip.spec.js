import React from "react";
import { render, cleanup } from "@testing-library/react";
import Tooltip from "../index";

afterEach(cleanup);

describe("Tooltip", () => {
  const Comp = (
    <Tooltip
      visible={true}
      children={[() => <button>Submit</button>, <h1>sdlkmsldkksm</h1>]}
    />
  );

  it("should take a snapshot", () => {
    const { asFragment } = render(Comp);

    expect(asFragment(Comp)).toMatchSnapshot();
  });

  // it("renders correctly", () => {
  //   const { queryByLabelText } = render(Comp);

  //   expect(queryByLabelText("Tooltip")).toBeTruthy();
  // });
});
