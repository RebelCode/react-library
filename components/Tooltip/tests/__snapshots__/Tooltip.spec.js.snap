// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should take a snapshot 1`] = `
<DocumentFragment>
  <button>
    Submit
  </button>
</DocumentFragment>
`;
